Threaded General User Objects

andrsd opened the pull request:
Allow creating threaded copies of general user objects 

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

andrsd committed:
Create threaded copies of FluidProperties objects

To make them thread safe when called from Materials, etc.

Refs #11731

andrsd committed:
Fixing #include guard

Should be all in CAPS

Refs #000

andrsd committed:
Threaded general user objects

Now, general users objects can be build per thread. User can set
"threaded" flag to inform MOOSE thay require threaded copies of their
user objects.

Closes #11731

andrsd committed:
Fixing validParams in ReadDoubleIndex class

Need to get params from GeneralUserObject not just UserObject.

andrsd commented:
@jasondhales Can you test this branch with your BISON changes to see if that fixes the threading problem you described in #11731? Thanks!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/210306/) on 01115f6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11834/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11834/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dd25ade99989c5f0075b570090833010a1df89af`

jasondhales commented:
If I change "threaded_copies" to "threaded" in FluidProperties.C, it works.

andrsd commented:
Thanks for catching that!

andrsd committed:
Fixing validParams in CrackDataSampler

Have to get validParams from your base class.

Refs #000

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/210566/) on c103549 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11834/site)

This comment will be updated on new commits.

andrsd committed:
Fixing validParams in GrainForcesPostprocessor

Have to get validParams from a base class.

Refs #000

friedmud commented on framework/src/loops/ComputeThreadedGeneralUserObjectsThread.C:
```diff
@@ -0,0 +1,52 @@
+    ParallelUniqueId puid;
+    _tid = bypass_threading ? 0 : puid.id;
+
+    const auto & objects = _user_objects.getActiveObjects();
```
I don't get it - won't this make every thread execute every object?

friedmud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -2955,6 +2971,18 @@ FEProblemBase::computeUserObjects(const ExecFlagType & type, const Moose::AuxGro
+  if (threaded_general.hasActiveObjects())
+  {
+    ComputeThreadedGeneralUserObjectsThread ctguot(*this, threaded_general);
+    Threads::parallel_reduce(EmptyRange(), ctguot);
```
This is what I was saying we shouldn't do.  You need to actually pass the range of userobjects here to ensure they all get executed.

friedmud reviewed (changes requested):
Some issues with execution

andrsd committed:
Prevent threaded general user objects with pthread

If a general user object used threads inside, we would have a threaded
region inside a threaded region and that would destroy people's machine.
We want to prevent that, so we only allow threaded general user objects
with OpenMP and TBB.

moosebuild commented:
Job [App tests](https://civet.inl.gov/job/210800/) on d732de0 : invalidated by @andrsd

grizzly is up-to-date

rwcarlsen commented:
This makes me sad - I will have to rewrite this code and resolve lots of conflicts :-/

jasondhales commented:
I'm not sure why this is an issue for @rwcarlsen, but I'm eager to see this merged.

friedmud reviewed (changes requested):
The code here looks good - but I would like to see one test of it.  Can you dream up something that would test this?

andrsd commented:
> The code here looks good - but I would like to see one test of it. Can you dream up something that would test this?

I flipped the "threaded" flag for the `FluidProperty` class, so this would be tested whenever we run tests in the `fluid_property` module. If that is not enough, I could build an user object that has some state, so that when it runs threaded and does not have the `threaded` flag on, it would produce a wrong result. If we get "lucky". Other than that, I am not sure how else to test this.

rwcarlsen commented:
It would be helpful if we have a reliable test (that fails without this
fix/feature) - if we don't I'm likely to introduce regressions soon when I
rework the code.

On Wed, Jul 18, 2018 at 2:44 PM, David Andrs <notifications@github.com>
wrote:

> The code here looks good - but I would like to see one test of it. Can you
> dream up something that would test this?
>
> I flipped the "threaded" flag for the FluidProperty class, so this would
> be tested whenever we run tests in the fluid_property module. If that is
> not enough, I could build an user object that has some state, so that when
> it runs threaded and does not have the threaded flag on, it would produce
> a wrong result. If we get "lucky". Other than that, I am not sure how else
> to test this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/11834#issuecomment-406068145>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABSeSqhK2YuGVjJdBata72T5ip-Pm-k1ks5uH54pgaJpZM4VL6Je>
> .
>

andrsd commented:
The problem is that the result would depend on the thread scheduler which we have no control over, so the test could sometimes pass even if there was something wrong :-(

friedmud commented:
The thread scheduler doesn't matter here - every threaded object will get
executed... even if not on separate threads.  All we need to verify is that
each one gets executed...

And, yes, I would like a reliable test in moose_test before this can go
in.  It's a tricky enough capability that we need to make sure it stays
working correctly.

On Wed, Jul 18, 2018 at 3:38 PM David Andrs <notifications@github.com>
wrote:

> The problem is that the result would depend on the thread scheduler which
> we have no control over, so the test could sometimes pass even if there was
> something wrong :-(
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/11834#issuecomment-406082690>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMfo1FVlhePg6xYhsMWv9QkRonJWaks5uH6rqgaJpZM4VL6Je>
> .
>

andrsd committed:
Adding tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/213922/) on 2986d2e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11834/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11834/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dd25ade99989c5f0075b570090833010a1df89af`

andrsd commented:
(Hopefully) reliable test added...

permcody commented:
ping @friedmud? 8 days no action

friedmud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -2955,6 +2972,22 @@ FEProblemBase::computeUserObjects(const ExecFlagType & type, const Moose::AuxGro
+      thread_ids[tid] = tid;
+
+    ComputeThreadedGeneralUserObjectsThread ctguot(*this, threaded_general);
+    Threads::parallel_reduce(ThreadIdRange(thread_ids.begin(), thread_ids.end()), ctguot);
```
Ok - the new comment is that this doesn't work with interdependencies.  Each ThreadedGeneralUserObject needs its own thread loop - then call `threadjoin()` and `finalize()` on it before starting the next one.

friedmud reviewed (changes requested):
Issue with interdependencies between these objects.

friedmud reviewed (approved):
👍

brianmoose commented:
Looks like our minimum clang testing has a problem with this PR: https://civet.inl.gov/job/218243/

andrsd commented:
#11961 should fix that
